src/pandas_profiling/model/dataframe_wrappers.py

```diff
@@ -0,0 +1,599 @@
+
+        Returns:
+
+        """
+        pass
+
+    @abstractmethod
+    def groupby_get_n_largest_dups(self, columns: List[str], n: int) -> pd.DataFrame:
+        """
+        get top n value counts of groupby count function
+
+        Args:
+            columns: columns to groupby on
+            n: top n value counts to return
+            for_duplicates: whether to only keep duplicate rows
+        Returns:
+
+        """
+        pass
+
+    @abstractmethod
+    def get_memory_usage(self, deep: bool = False) -> pd.Series:
+        """
+        Get memory usage of dataframe
+
+        Args:
+            deep: * For Pandas - interrogating object dtypes for system-level memory consumption
+
+        Returns: A Pandas Series whose index is the original column names and whose values is the
+         memory usage of each column in bytes.
+
+        """
+        pass
+
+    @abstractmethod
+    def head(self, n: int):
+        """
+        Get top n rows of dataframe.
+        Only call within singledispatch methods
+
+        Args:
+            n: top n rows
+
+        Returns: dataframe with only top n rows
+
+        """
+        pass
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """
+        Get the number of rows in a dataframe as an int
+
+        Returns: number of rows in column
+
+        """
+        pass
+
+    @abstractmethod
+    def __getitem__(self, key):
+        pass
+
+
+class PandasDataFrame(GenericDataFrame):
+    """
+    This class is the abstract layer for the pandas dataframe.
+
+    It implements the check_if_corresponding_engine function to check if an obj is a pandas dataframe and can be wrapped.
+
+    It implements the get_pandas_df method to enable singledispatch wrapped functions to
+    operate on the internal pandas dataframe without breaking abstraction
+
+    """
+
+    engine = "pandas"
+
+    def __init__(self, df):
+        super().__init__()
+        # self.df holds the underlying data object
+        self.df = df
+        self.n_rows = len(self.df)
+
+    @staticmethod
+    def check_if_corresponding_engine(obj) -> bool:
+        """
+        Check if pandas dataframe using isinstance. More pythonic way of checking as opposed to spark type check.
+
+        Args:
+            obj: check if object is a pandas dataframe
+
+        Returns: True if the __module__ and __name__ of object matches pandas dataframe, else false
+        """
+        return isinstance(obj, pd.DataFrame)
+
+    @staticmethod
+    def preprocess(df: pd.DataFrame) -> pd.DataFrame:
+        """Preprocess the dataframe
+
+        - Appends the index to the dataframe when it contains information
+        - Rename the "index" column to "df_index", if exists
+        - Convert the DataFrame's columns to str
+
+        Args:
+            df: the pandas DataFrame
+
+        Returns:
+            The preprocessed DataFrame
+        """
+
+        # Treat index as any other column
+        if (
```

Comment:
I think this is getting at whether the index is a `RangeIndex` with start=0, and step=1 (the default index type) which can be checked without building a new index and comparing all values directly. e.g.

```
if not isinstance(df.index, pd.RangeIndex) or not (df.index.start == 0 and df.index.step == 1):
  df = df.reset_index()
```

That conditional looks a little weird but it's equivalent to `not (isinstance(df.index, pd.RangeIndex) and (df.index.start == 0 and df.index.step == 1))` via De Morgan's law.
```